src/lib/pubkey/xmss/xmss_index_registry.h

```diff
@@ -0,0 +1,109 @@
+/**
+ * XMSS Index Registry
+ * (C) 2016 Matthias Gierlings
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ **/
+
+#ifndef BOTAN_XMSS_INDEX_REGISTRY_H__
+#define BOTAN_XMSS_INDEX_REGISTRY_H__
+
+#include <stdint.h>
+#include <cstddef>
+#include <limits>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <botan/hash.h>
+#include <botan/secmem.h>
+#include <botan/types.h>
+#include <botan/atomic.h>
+
+namespace Botan {
+
+/**
+ * A registry for XMSS private keys, keeps track of the leaf index for
+ * independend copies of the same key.
+ **/
```

Comment:
Actually this is not defined in the standard. The standard only states that reuse of one time keys must be prevented. Since the XMSS_Private_Key is copyable, there is a chance that two independent copies of the same private key exist within one application. The leaf indices might be updated independently which would likely result in key reuse. The purpose of the index registry is to map all copies of a private key to a shared leaf index. That way even if a library user accidentally copies the same key and uses multiple instances for signing, the leaf index will still be properly updated across all instances of the same key.